src/aerospike_clj/client.clj

```diff
@@ -57,24 +59,42 @@
 
 (defn init-simple-aerospike-client
   "hosts should be a seq of known hosts to bootstrap from. Optional conf map _can_ have:
-  :event-loops - a client compatible event loop instace
+  `:event-loops` - a client compatible event loop instace
   client policy configuration keys (see policy/create-client-policy)
-  :client-policy - a ready ClientPolicy
-  \"username\"
-  :port (default is 3000)
-  :client-events an implementation of ClientEvents"
+  `:client-policy` - a ready ClientPolicy
+  \"username\" (may be `nil`)
+  `:port` (default is 3000)
+  `:client-events` an implementation of ClientEvents
+  \"maxCommandsInProcess\" should be more than 5 (default `nil`. used by ClientPolicy in EventLoops)
+  \"maxCommandsInQueue\" should be more than 0 (default `nil`. used by ClientPolicy in EventLoops)
+  `block-on-queue-full?` a boolean to enable back-pressure if maxCommandsInProcess and 
+  maxCommandsInQueue are enabled." 
   ([hosts aero-ns]
    (init-simple-aerospike-client hosts aero-ns {}))
   ([hosts aero-ns conf]
    (let [cluster-name (utils/cluster-name hosts)
-         event-loops (:event-loops conf (create-event-loops conf))
-         client-policy (:client-policy conf (policy/create-client-policy event-loops conf))]
+         ^EventLoops event-loops (:event-loops conf (create-event-loops conf))
+         event-loops-size (.getSize event-loops)
+         client-policy (:client-policy conf (policy/create-client-policy event-loops conf))
+         block-on-queue-full? (:block-on-queue-full? conf false)
+         ^Semaphore commands-sem (when (and block-on-queue-full?
+                                            (pos? (get conf "maxCommandsInProcess" 0))
+                                            (pos? (get conf "maxCommandsInQueue" 0)))
+                                   (let [permits (+ (* event-loops-size
+                                                       (get conf "maxCommandsInProcess"))
+                                                    (get conf "maxCommandsInQueue"))]
+                                     (Semaphore. permits true)))]
      (println (format ";; Starting aerospike clients for clusters %s with username %s" cluster-name (get conf "username")))
      (map->SimpleAerospikeClient {:ac (create-client hosts client-policy (:port conf 3000))
                                   :el event-loops
                                   :dbns aero-ns
                                   :cluster-name cluster-name
-                                  :client-events (:client-events conf)}))))
+                                  :client-events (:client-events conf)
+                                  :commands-sem commands-sem}))))
+
+(defn- acquire-permit [client]
+  (when-let [sem (:commands-sem client)]
+    (.acquire ^Semaphore sem)))
```

Comment:
should also have a release-permit as there is a lot of code repetition.
both functions should get either the client or the semaphore itself:
```clojure
(defn- release-permit [sem]
  (when sem
    (.release ^Semaphore sem)))
```